Error: Cannot find module 'env-var'

# Description

#176: change of env to env-var causes failure to load. Is "env-var" now a dependency?

### Log

```
[2020-10-04 08:04:25.261] [exthost] [error] Activating extension cschlosser.doxdocgen failed due to an error:
[2020-10-04 08:04:25.265] [exthost] [error] Error: Cannot find module 'env-var'
Require stack:
- c:\Users\user\.vscode\extensions\cschlosser.doxdocgen-0.8.0\out\Lang\Cpp\CppDocGen.js
- c:\Users\user\.vscode\extensions\cschlosser.doxdocgen-0.8.0\out\Lang\Cpp\CppParser.js
- c:\Users\user\.vscode\extensions\cschlosser.doxdocgen-0.8.0\out\CodeParserController.js
- c:\Users\user\.vscode\extensions\cschlosser.doxdocgen-0.8.0\out\extension.js
- c:\Users\user\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\loader.js
- c:\Users\user\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-amd.js
- c:\Users\user\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js```

--- a/.travis.yml
+++ b/.travis.yml
@@ -42,8 +42,8 @@ jobs:
       os: osx
     - stage: release
       script:
-        - npm install vsce@1.31.0
-        - vsce package
+        - npm install vsce
+        - vsce package --no-yarn
       os: linux
       deploy:
         - provider: releases
@@ -53,7 +53,7 @@ jobs:
           on:
             tags: true
         - provider: script
-          script: vsce publish -p $VSMARKETPLACE_ACCESS_TOKEN
+          script: vsce publish --no-yarn -p $VSMARKETPLACE_ACCESS_TOKEN
           skip_cleanup: true
           on:
             tags: true

--- a/.travis.yml
+++ b/.travis.yml
@@ -44,6 +44,7 @@ jobs:
       script:
         - npm install vsce
         - vsce package --no-yarn
+        - ls -alh
       os: linux
       deploy:
         - provider: releases

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,15 @@
 # Change Log
 
+## [1.0.0]
+
+### Revert
+
+- Revert reverting Replace environment variables in templated strings. If no environment variable can be found the name of the variable will be inserted (#110)
+
+### Other
+
+- vsce packaging is forcing yarn even if no yarn config exists. Override this behavior now to use npm.
+
 ## [0.8.2]
 
 ### Other

--- a/package.json
+++ b/package.json
@@ -2,7 +2,7 @@
   "name": "doxdocgen",
   "displayName": "Doxygen Documentation Generator",
   "description": "Let me generate Doxygen documentation from your source code for you.",
-  "version": "0.8.2",
+  "version": "1.0.0",
   "publisher": "cschlosser",
   "engines": {
     "vscode": "^1.37.0"

--- a/package.json
+++ b/package.json
@@ -253,6 +253,7 @@
     "test": "npm run compile && node ./node_modules/vscode/bin/test"
   },
   "dependencies": {
+    "env-var": "^4.1.0",
     "moment": "^2.20.1",
     "opn": "^5.2.0"
   },

--- a/src/Lang/Cpp/CppDocGen.ts
+++ b/src/Lang/Cpp/CppDocGen.ts
@@ -1,10 +1,10 @@
+import * as env from "env-var";
 import * as moment from "moment";
-import { Position, Range, Selection, TextEditor, TextLine, WorkspaceEdit } from "vscode";
+import { Position, Range, Selection, TextEditor } from "vscode";
 import { IDocGen } from "../../Common/IDocGen";
 import { Config } from "../../Config";
 import { CppArgument } from "./CppArgument";
 import * as CppParser from "./CppParser";
-import { CppParseTree } from "./CppParseTree";
 import { CppToken, CppTokenType } from "./CppToken";
 
 export enum SpecialCase {
@@ -144,9 +144,31 @@ export class CppDocGen implements IDocGen {
         return indentedString;
     }
 
+    protected getEnvVars(replace: string): string {
+        let replacement = replace;
+        const regex = /\$\{env\:([\w|\d|_]+)\}/m;
+        let match: RegExpExecArray;
+
+        // tslint:disable-next-line:no-conditional-assignment
+        while ((match = regex.exec(replacement)) !== null) {
+            if (match.index === regex.lastIndex) {
+                regex.lastIndex++;
+            }
+
+            const m = match[1];
+
+            const envVar: string = env.get(m, m).asString();
+
+            replacement = replacement.replace("${env:" + m + "}", envVar);
+        }
+
+        return replacement;
+    }
+
     protected getTemplatedString(replace: string, template: string, param: string): string {
         const replacedTemplate = template.replace(replace, param);
-        return this.getIndentedTemplate(replacedTemplate);
+        const replacedWithEnv = this.getEnvVars(replacedTemplate);
+        return this.getIndentedTemplate(replacedWithEnv);
     }
 
     protected getMultiTemplatedString(replace: string[], template: string, param: string[]): string {
@@ -156,7 +178,7 @@ export class CppDocGen implements IDocGen {
               template = template.replace(replace[i], param[i]);
             }
         }
-        return template;
+        return this.getEnvVars(template);
     }
 
     protected getSmartText(): string {
@@ -453,13 +475,15 @@ export class CppDocGen implements IDocGen {
                             lines,
                             this.cfg.typeTemplateReplace,
                             this.cfg.Generic.returnTemplate,
-                            returnParams
+                            returnParams,
                         );
                     }
                     break;
                 }
                 case "custom": {
-                    lines.push(...this.cfg.Generic.customTags);
+                    this.cfg.Generic.customTags.forEach((elem) => {
+                        lines.push(this.getEnvVars(elem));
+                    });
                     break;
                 }
                 default: {

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -5,8 +5,8 @@ import * as vscode from "vscode";
 import CodeParserController from "./CodeParserController";
 
 enum Version {
-    CURRENT = "0.8.2",
-    PREVIOUS = "0.8.1",
+    CURRENT = "1.0.0",
+    PREVIOUS = "0.8.2",
     KEY = "doxdocgen_version",
 }
 

--- a/src/test/CppTests/Config.test.ts
+++ b/src/test/CppTests/Config.test.ts
@@ -227,4 +227,24 @@ suite("C++ - Configuration Tests", () => {
         assert.equal("/**\n * @note\n */", result);
     });
 
+    test("Env variable", () => {
+        testSetup.cfg = new Config();
+        testSetup.cfg.Generic.order = ["custom"];
+        if (process.platform === "win32") {
+            testSetup.cfg.Generic.customTags = ["@author ${env:USERNAME}"];
+            const res = testSetup.SetLine("void foo();").GetResult();
+            // USERNAME env var is different for everybody
+            assert.notEqual("/**\n * @author USERNAME\n */", res);
+        } else {
+            testSetup.cfg.Generic.customTags = ["@author ${env:USER}"];
+            const res = testSetup.SetLine("void foo();").GetResult();
+            // USER env var is different for everybody
+            assert.notEqual("/**\n * @author USER\n */", res);
+        }
+
+        testSetup.cfg.Generic.customTags = ["@author ${env:MY_VARIABLE}"];
+        const result = testSetup.SetLine("void foo();").GetResult();
+        assert.equal("/**\n * @author MY_VARIABLE\n */", result);
+    });
+
 });

